src/forcing/ForcingsEngine.cpp

```diff
@@ -0,0 +1,239 @@
+    const auto var_dim  = var_outputs_.size();
+
+    bmi_->Update();
+    time_current_index_++;
+    
+    const auto* ptr = static_cast<int*>(bmi_->GetValuePtr("CAT-ID"));
+    var_divides_ = std::vector<int>(ptr, ptr + id_dim);
+    var_cache_ = decltype(var_cache_){{ time_dim, id_dim, var_dim }};
+
+    // Cache initial iteration
+    const auto size = var_divides_.size();
+    for (size_type vi = 0; vi < var_outputs_.size(); vi++) {
+        // Get current values for each variable
+        const auto var_span = boost::span<const double>{
+            static_cast<double*>(bmi_->GetValuePtr(var_outputs_[vi])),
+            size
+        };
+
+        for (size_type di = 0; di < size; di++) {
+            // Set current variable value for each divide
+            var_cache_.at({{time_current_index_, di, vi}}) = var_span[di];
+        }
+    }
+
+    time_current_index_++;
+};
+
+ForcingsEngine::~ForcingsEngine()
+{
+    finalize();
+}
+
+ForcingsEngine& ForcingsEngine::instance(
+    const std::string& init,
+    std::size_t time_start,
+    std::size_t time_end
+)
+{
+    if (instances_[init] == nullptr) {
+        instances_[init] = std::make_unique<ForcingsEngine>(init, time_start, time_end);
+    }
+
+    return *instances_[init];
+}
+
+void ForcingsEngine::check_runtime_requirements()
+{
+    
+    // Check that the python module is installed.
+    {
+        auto interpreter_ = utils::ngenPy::InterpreterUtil::getInstance();
+        try {
+            auto mod = interpreter_->getModule("NextGen_Forcings_Engine");
+            auto cls = mod.attr("BMIForcingsEngine").cast<py::object>();
+        } catch(std::exception& e) {
+            throw std::runtime_error{
+                "Failed to initialize ForcingsEngine: ForcingsEngine python module is not installed or is not properly configured. (" + std::string{e.what()} + ")"
+            };
+        }
+    }
+
+    // Check that the WGRIB2 environment variable is defined
+    {
+        const auto* wgrib2_exec = std::getenv("WGRIB2");
+        if (wgrib2_exec == nullptr) {
+            throw std::runtime_error{"Failed to initialize ForcingsEngine: $WGRIB2 is not defined"};
+        }
+    }
+}
+
+boost::span<const std::string> ForcingsEngine::outputs() const noexcept
+{
+    return var_outputs_;
+}
+
+auto ForcingsEngine::time_begin() const noexcept
+  -> const clock_type::time_point&
+{
+    return time_start_;
+}
+
+auto ForcingsEngine::time_end() const noexcept
+  -> const clock_type::time_point&
+{
+    return time_end_;
+}
+
+auto ForcingsEngine::time_step() const noexcept
+  -> const clock_type::duration&
+{
+    return time_step_;
+}
+
+auto ForcingsEngine::time_index(time_t ctime) const noexcept
+  -> size_type
+{
+    const auto epoch = clock_type::from_time_t(ctime);
+
+    if (epoch < time_begin() || epoch > time_end()) {
+        return bad_index;
+    }
+
+    const auto offset = (epoch - time_begin()) / time_step();
```

Comment:
Do we need to worry about rounding and loss of precision issues with this calculation? It's not clear to me whether the `time_point` representation is integral or floating point, since cppreference just says "signed arithmetic type"

Reply:
I _think_ (heavy emphasis on think) that this is an integral operation. At the very least, `decltype(epoch - time_begin())` evaluates to a `chrono::duration<long, std::ratio<...>>`, and the division of that against `decltype(time_step())` has a common type of `long`.